From 03cf937ce263ae207cedb2fb231c9f67c6746981 Mon Sep 17 00:00:00 2001
From: Hiemanshu Sharma <hiemanshu@gmail.com>
Date: Wed, 24 Jun 2020 00:55:53 +0530
Subject: [PATCH] Add tags to campaign exports

---
 src/workers/jobs.js | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/src/workers/jobs.js b/src/workers/jobs.js
index 5770cfb0..195c3aea 100644
--- a/src/workers/jobs.js
+++ b/src/workers/jobs.js
@@ -736,6 +736,11 @@ export async function exportCampaign(job) {
     convertedMessages = await Promise.all(convertedMessages);
     finalCampaignMessages = finalCampaignMessages.concat(convertedMessages);
     let convertedContacts = contacts.map(async contact => {
+      const tags = await r
+        .knex("tag_campaign_contact")
+        .where("campaign_contact_id", contact.id)
+        .leftJoin("tag", "tag.id", "tag_campaign_contact.tag_id");
+
       const contactRow = {
         campaignId: campaign.id,
         campaign: campaign.title,
@@ -756,7 +761,8 @@ export async function exportCampaign(job) {
           : "false",
         "contact[messageStatus]": contact.message_status,
         "contact[errorCode]": contact.error_code,
-        "contact[external_id]": contact.external_id
+        "contact[external_id]": contact.external_id,
+        "contact[tags]": tags.length > 0 ? tags.map(tag => tag.name) : null
       };
       const customFields = JSON.parse(contact.custom_fields);
       Object.keys(customFields).forEach(fieldName => {
-- 
2.24.3 (Apple Git-128)

